sw/device/lib/dif/dif_pwrmgr.c

```diff
@@ -4,8 +4,543 @@
+    dif_pwrmgr_request_sources_t *sources) {
+  if (pwrmgr == NULL || !is_valid_req_type(req_type) || sources == NULL) {
+    return kDifPwrmgrBadArg;
+  }
+
+  request_reg_info_t reg_info = request_reg_infos[req_type];
+  uint32_t reg_val = mmio_region_read32(pwrmgr->params.base_addr,
+                                        reg_info.current_sources_reg_offset);
+  *sources = bitfield_field32_read(reg_val, reg_info.bitfield);
+
+  return kDifPwrmgrOk;
+}
+
+dif_pwrmgr_result_t dif_pwrmgr_request_sources_lock(
+    const dif_pwrmgr_t *pwrmgr, dif_pwrmgr_req_type_t req_type) {
+  if (pwrmgr == NULL || !is_valid_req_type(req_type)) {
+    return kDifPwrmgrBadArg;
+  }
+
+  // Only a single bit of this register is significant, thus we don't perform a
+  // read-modify-write. Setting this bit to 0 locks sources.
+  mmio_region_write32(pwrmgr->params.base_addr,
+                      request_reg_infos[req_type].write_enable_reg_offset, 0);
+
+  return kDifPwrmgrOk;
+}
+
+dif_pwrmgr_result_t dif_pwrmgr_request_sources_is_locked(
+    const dif_pwrmgr_t *pwrmgr, dif_pwrmgr_req_type_t req_type,
+    bool *is_locked) {
+  if (pwrmgr == NULL || !is_valid_req_type(req_type) || is_locked == NULL) {
+    return kDifPwrmgrBadArg;
+  }
+
+  *is_locked = request_sources_is_locked(pwrmgr, req_type);
+
+  return kDifPwrmgrOk;
+}
+
+dif_pwrmgr_result_t dif_pwrmgr_wakeup_request_recording_set_enabled(
+    const dif_pwrmgr_t *pwrmgr, dif_pwrmgr_toggle_t new_state) {
+  if (pwrmgr == NULL || !is_valid_toggle(new_state)) {
+    return kDifPwrmgrBadArg;
+  }
+
+  // Only a single bit of this register is significant, thus we don't perform a
+  // read-modify-write. Setting this bit to 1 disables recording.
+  uint32_t reg_val =
+      bitfield_bit32_write(0, PWRMGR_WAKE_INFO_CAPTURE_DIS_VAL,
+                           new_state == kDifPwrmgrToggleDisabled);
+  mmio_region_write32(pwrmgr->params.base_addr,
+                      PWRMGR_WAKE_INFO_CAPTURE_DIS_REG_OFFSET, reg_val);
+
+  return kDifPwrmgrOk;
+}
+
+dif_pwrmgr_result_t dif_pwrmgr_wakeup_request_recording_get_enabled(
+    const dif_pwrmgr_t *pwrmgr, dif_pwrmgr_toggle_t *cur_state) {
+  if (pwrmgr == NULL || cur_state == NULL) {
+    return kDifPwrmgrBadArg;
+  }
+
+  uint32_t reg_val = mmio_region_read32(
+      pwrmgr->params.base_addr, PWRMGR_WAKE_INFO_CAPTURE_DIS_REG_OFFSET);
+  // Recoding is disabled if this bit is set to 1.
+  *cur_state = bitfield_bit32_read(reg_val, PWRMGR_WAKE_INFO_CAPTURE_DIS_VAL)
+                   ? kDifPwrmgrToggleDisabled
+                   : kDifPwrmgrToggleEnabled;
+
+  return kDifPwrmgrOk;
+}
+
+dif_pwrmgr_result_t dif_pwrmgr_wakeup_reason_get(
+    const dif_pwrmgr_t *pwrmgr, dif_pwrmgr_wakeup_reason_t *reason) {
+  if (pwrmgr == NULL || reason == NULL) {
+    return kDifPwrmgrBadArg;
+  }
+
+  uint32_t reg_val =
+      mmio_region_read32(pwrmgr->params.base_addr, PWRMGR_WAKE_INFO_REG_OFFSET);
+  // Wakeup types.
```

Comment:
What's this comment for?

Reply:
Pretty terse :) For the first field of the struct.